consensus/actor_state.go

```diff
@@ -0,0 +1,96 @@
+	"github.com/filecoin-project/go-filecoin/address"
+	"github.com/filecoin-project/go-filecoin/state"
+	"github.com/filecoin-project/go-filecoin/types"
+	"github.com/filecoin-project/go-filecoin/vm"
+)
+
+// Abstracts over a store of blockchain state.
+type chainStateChainReader interface {
+	GetHead() types.TipSetKey
+	GetTipSetState(context.Context, types.TipSetKey) (state.Tree, error)
+	GetTipSet(types.TipSetKey) (types.TipSet, error)
+}
+
+// ActorState knows how to send read-only messages for querying actor state.
+type ActorState struct {
+	// To get the head tipset state root.
+	chainReader chainStateChainReader
+	// To load the tree for the head tipset state root.
+	cst *hamt.CborIpldStore
+	// For vm storage.
+	bs bstore.Blockstore
+}
+
+// NewActorState constructs a ActorState.
+func NewActorState(chainReader chainStateChainReader, cst *hamt.CborIpldStore, bs bstore.Blockstore) *ActorState {
+	return &ActorState{chainReader, cst, bs}
+}
+
+type ActorStateQueryer interface {
+	Query(ctx context.Context, optFrom, to address.Address, method string, params ...interface{}) ([][]byte, error)
+}
+
+// Queryer returns a query interface to query the chain at a particular tipset
+func (cs ActorState) Queryer(ctx context.Context, baseKey types.TipSetKey) (ActorStateQueryer, error) {
+	st, err := cs.chainReader.GetTipSetState(ctx, baseKey)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to load tree for the state root of tipset: %s", baseKey.String())
+	}
+	base, err := cs.chainReader.GetTipSet(baseKey)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get tipset: %s", baseKey.String())
+	}
+	h, err := base.Height()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get the head tipset height")
+	}
+
+	return cs.stateTreeQueryer(st, types.NewBlockHeight(h)), nil
+}
+
+// stateTreeQueryer returns a query interface to query the chain for a particular state tree and optional block height
+func (cs ActorState) stateTreeQueryer(st state.Tree, bh *types.BlockHeight) ActorStateQueryer {
+	return NewProcessorQueryer(st, vm.NewStorageMap(cs.bs), bh)
+}
+
+// ProcessorQueryer queries the chain at a particular tipset
+type ProcessorQueryer struct {
+	st     state.Tree
+	vms    vm.StorageMap
+	height *types.BlockHeight
+}
+
+// NewProcessorQueryer creates an ActorStateQueryer
+func NewProcessorQueryer(st state.Tree, vms vm.StorageMap, height *types.BlockHeight) ActorStateQueryer {
+	return &ProcessorQueryer{
+		st:     st,
+		vms:    vms,
+		height: height,
+	}
+}
+
+// Query sends a read-only message against the state of the provided base tipset.
+func (q *ProcessorQueryer) Query(ctx context.Context, optFrom, to address.Address, method string, params ...interface{}) ([][]byte, error) {
+	encodedParams, err := abi.ToEncodedValues(params...)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to encode message params")
+	}
+
+	r, ec, err := CallQueryMethod(ctx, q.st, q.vms, to, method, encodedParams, optFrom, q.height)
```

Comment:
Highly doubtful this belongs in this PR (or is even the right call) but it seems like putting `CallQueryMethod` on a processor object and having the *Processor*Queryer keep track of a processor internally is a good direction.

You might imagine the Processor keeps track of the history-dependent vms and new processing invocations take in the per-call state and height params which the `Queryer` could continue to track.

Reply:
More future design thoughts:

This also makes me think that using a variant of the `Queryer` (maybe even an unmodified `Queryer`) within the actual block processing code could make sense.  I.e. batched message application for processing blocks and tipsets could be an operation that is a method on a thing that takes in a `Queryer` and then uses the `Queryer` mutliple times to actually do the message application (along with block reward state mods etc).

This would potentially make the many layers of redirection within the processor `ProcessTipSet` --> `ProcessBlock` --> `ApplyMessagesAndPayRewards` --> `ApplyMessage` --> `attemptApplyMessage` more explicit and obvious by moving the redirection to specific types.  For example we could have a `BlockProcessor` that owns a queryer and logic for processing tipsets and blocks, a `Queryer` as defined here or slightly varied which manages running a single message through the state machine, and a `StateMachineProcessor`/`StateMachineRunner` that keeps the logic for `attemptApplyMessage` or that can be easily mocked and passed to the `Queryer`

Reply:
Also maybe the `Queryer` becomes the `StateMachineRunner` above, i.e. it makes more sense to group the `attemptApplyMessage` logic in directly and mock out the queryer than add another layer of indirection.  I think the clearest thing I can see is that running a message through the state machine and handling the block/tipset processing logic definitely deserve separate owning abstractions.

Reply:
Much of the the motivation of this PR comes from the fact that `CallQueryMethod` is a free function in the `consensus` package that just happens to live in the `processor.go` file. Otherwise, giving `ProcessorQueryer` a processor seems like a good idea.

My understanding of the current state is that `Expected` runs answers questions about chain state including chain weight and the validity of tip sets. `Processor` concerns itself with processing signed messages in various packages (blocks, slices, single) for state transitions or simply to query. The actual execution of a message is handled in the `vm` package. 

I think I sort of get where you're going with the refactoring ideas. Perhaps instead of a `Queryer`, we have a better `Processor` (Transitioner?) abstraction that is tied to some base state. The caller will be responsible for retrieving a `Processesor` with the correct state. They could then use the Processor to `play` messages in tip sets, blocks or individually. Conceivably the `Processor` would return a new `Processor` in addition to the results that would represent the new state after the messages had been played, so the processor's internal state tree would not need to mutate.